trino-aws-proxy/src/main/java/io/trino/aws/proxy/server/rest/TrinoLogsResource.java

```diff
@@ -0,0 +1,242 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.aws.proxy.server.rest;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.common.base.Splitter;
+import com.google.common.collect.ImmutableSet;
+import com.google.inject.Inject;
+import io.trino.aws.proxy.server.rest.RequestLoggerController.SaveEntry;
+import io.trino.aws.proxy.server.rest.TrinoLogsResource.GetLogEventsResponse.Event;
+import io.trino.aws.proxy.spi.rest.Request;
+import jakarta.ws.rs.HeaderParam;
+import jakarta.ws.rs.POST;
+import jakarta.ws.rs.Produces;
+import jakarta.ws.rs.WebApplicationException;
+import jakarta.ws.rs.core.Context;
+import jakarta.ws.rs.core.MediaType;
+import jakarta.ws.rs.core.Response;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Locale;
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Predicate;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.aws.proxy.spi.signing.SigningServiceType.S3;
+import static io.trino.aws.proxy.spi.signing.SigningServiceType.STS;
+import static jakarta.ws.rs.core.Response.Status.BAD_REQUEST;
+import static java.util.Objects.requireNonNull;
+
+public class TrinoLogsResource
+{
+    private static final Set<String> DEFAULT_STREAMS = ImmutableSet.of(S3.serviceName(), STS.serviceName());
+
+    private enum TokenType
+    {
+        BACKWARDS("b/"),
+        FORWARDS("f/");
+
+        private final String prefix;
+
+        private Optional<String> atIndex(int index)
+        {
+            return Optional.of(prefix + index);
+        }
+
+        TokenType(String prefix)
+        {
+            this.prefix = requireNonNull(prefix, "prefix is null");
+        }
+    }
+
+    private final RequestLoggerController loggerController;
+    private final ObjectMapper objectMapper;
+
+    // see https://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_GetLogEvents.html
+    public record GetLogEventsRequest(
+            Optional<Long> startTime,
+            Optional<Long> endTime,
+            Optional<Integer> limit,
+            Optional<String> nextToken,
+            Optional<Boolean> startFromHead,
+            Optional<Set<String>> logStreamNames,
+            Optional<String> logStreamName)
+    {
+        public GetLogEventsRequest
+        {
+            requireNonNull(startTime, "startTime is null");
+            requireNonNull(endTime, "endTime is null");
+            requireNonNull(limit, "limit is null");
+            requireNonNull(nextToken, "nextToken is null");
+            requireNonNull(startFromHead, "startFromHead is null");
+            requireNonNull(logStreamNames, "logStreamNames is null");
+            requireNonNull(logStreamName, "logStreamName is null");
+
+            if (logStreamNames.isEmpty() && logStreamName.isEmpty()) {
```

Comment:
Depending on how the objectmapper is configured, `logStreamNames.isEmpty()` may not be reliable right? We sometimes default collections to empty ones, meaning this could be `Optional.of(Set.of())`

Should we check whether `logStreamNames` is an empty optional or an empty set?
Or alternatively just make it a set and treat an empty set as being missing in the request

Reply:
We're injecting the ObjectMapper in the main module which has the `addNullCollectionModule()`. So, it will always be consistent. Maybe if this was in the SPI module.